Objects/genobject.c

```diff
@@ -368,8 +385,26 @@ _PyGen_yf(PyGenObject *gen)
     return NULL;
 }
 
+PyObject *
+_PyGen_yf(PyGenObject *gen)
+{
+    PyObject *res;
+    Py_BEGIN_CRITICAL_SECTION(gen);
+    res = _PyGen_yf_lock_held(gen);
+    Py_END_CRITICAL_SECTION();
+    return res;
+}
+
+/*[clinic input]
+@critical_section
+generator.close as gen_close
```

Comment:
Would it be possible to use a different name for the Argument Clinic function?

```suggestion
generator.close as gen_close_meth
```

And have a separated `int gen_close(PyGenObject *gen)` function which returns -1 on error or 0 on success.